### PR TITLE
feat(lambda): add grantInvokeCompositePrincipal method

### LINKED ISSUE #27503: aws-lambda: Invalid principal type for Lambda permission statement: CompositePrincipal.
### Describe the bug

I am getting an error when trying to use a CompositePrincipal to `grantInvoke` permissions on a lambda.

### Expected Behavior

I expect it to work and allow all of the account principals to invoke the lambda.

### Current Behavior

Currently it errors out and says it is unsupported with the following error:

```
Error: Invalid principal type for Lambda permission statement: CompositePrincipal. Supported: AccountPrincipal, ArnPrincipal, ServicePrincipal, OrganizationPrincipal
```

### Reproduction Steps

This code results in the error:

```typescript
const accounts = new iam.CompositePrincipal(
  new iam.AccountPrincipal('1234'),
  new iam.AccountPrincipal('5678'),
);
myLambda.grantInvoke(accounts);
```

### Possible Solution

_No response_

### Additional Information/Context

Possibly related:
- https://github.com/aws/aws-cdk/issues/3264
- https://github.com/aws/aws-cdk/pull/3501

### CDK CLI Version

2.95.0 (build cfa7e88)

### Framework Version

_No response_

### Node.js Version

v18.17.1

### OS

macOS Sonoma

### Language

TypeScript

### Language Version

TypeScript 5.2

### Other information

_No response_

### PR DESCRIPTION
**Pull request by msambol**
Add ability to grant function access to `CompositePrincipal`.

Closes #27503.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-iam/lib/principals.ts [hunk 1]
```diff
@@ -825,7 +825,7 @@ export class StarPrincipal extends PrincipalBase {
  */
 export class CompositePrincipal extends PrincipalBase {
   public readonly assumeRoleAction: string;
-  private readonly principals = new Array<IPrincipal>();
+  public readonly principals = new Array<IPrincipal>();
```
Comment (msambol) [inline packages/aws-cdk-lib/aws-iam/lib/principals.ts:828]: Is this OK to make public?

File: packages/aws-cdk-lib/aws-lambda/README.md [hunk 2]
```diff
@@ -249,6 +249,18 @@ const servicePrincipalWithConditions = servicePrincipal.withConditions({
 fn.grantInvoke(servicePrincipalWithConditions);
 ```
 
+### Grant function access to CompositePrincipal
+
+```ts
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-lambda/README.md:254]:
```suggestion
### Grant function access to CompositePrincipal

To grant invoke permissions to a `CompositePrincipal` use the `grantInvokeCompositePrincipal` method:

```ts
```
Short description.

File: packages/aws-cdk-lib/aws-lambda/lib/function-base.ts [hunk 3]
```diff
@@ -449,6 +454,19 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
...
+   */
+  public grantInvokeCompositePrincipal(compositePrincipals: iam.CompositePrincipal): iam.Grant[] {
+    let grants: iam.Grant[] = [];
+
+    compositePrincipals.principals.forEach((principal) => {
+      grants.push(this.grantInvoke(principal));
+    });
+
+    return grants;
+  }
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-lambda/lib/function-base.ts:468]:
```suggestion
  public grantInvokeCompositePrincipal(compositePrincipals: iam.CompositePrincipal): iam.Grant[] {
    return compositePrincipals.principals.map((principal) => this.grantInvoke(principal));
  }
```
Cleaner.

File: packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.ts [hunk 4]
```diff
@@ -16,3 +16,8 @@ const fn = new lambda.Function(stack, 'MyLambda', {
 fn.grantInvoke(new iam.AnyPrincipal().inOrganization('o-yyyyyyyyyy'));
 
 fn.grantInvoke(new iam.OrganizationPrincipal('o-xxxxxxxxxx'));
+
```
Comment (lpizzinidev) [inline packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.ts:19]: Can you please add a test case also for the `grantInvokeUrl` method?

File: packages/aws-cdk-lib/aws-lambda/README.md [hunk 5]
```diff
@@ -249,6 +249,18 @@ const servicePrincipalWithConditions = servicePrincipal.withConditions({
...
+declare const fn: lambda.Function;
+const compositePrincipal = new iam.CompositePrincipal(
+  new iam.OrganizationPrincipal('o-zzzzzzzzzz'),
+  new iam.ServicePrincipal('apigateway.amazonaws.com'),
+)
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-lambda/README.md:259]:
```suggestion
);
```
Semicolon.

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Looks good overall 👍
Just some small adjustments to be made in my opinion.

**Comment by msambol** [comment]
@lpizzinidev Updated per your feedback—thanks!

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍

File: packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.js.snapshot/lambda-permissions.assets.json [hunk 6]
```diff
@@ -1,15 +1,15 @@
 {
   "version": "34.0.0",
```
Comment (vinayak-kukreja) [inline packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.js.snapshot/lambda-permissions.assets.json:2]: Hey, I see the version here was not updated. Just want to confirm if these tests were dry-run? If so, please let me know and I can run these for this PR

File: packages/aws-cdk-lib/aws-cloudfront/lib/experimental/edge-function.ts [hunk 7]
```diff
@@ -124,6 +124,9 @@ export class EdgeFunction extends Resource implements lambda.IVersion {
   public grantInvokeUrl(identity: iam.IGrantable): iam.Grant {
     return this.lambda.grantInvokeUrl(identity);
   }
+  public grantInvokeCompositePrincipal(compositePrincipals: iam.CompositePrincipal): iam.Grant[] {
```
Comment (vinayak-kukreja) [inline packages/aws-cdk-lib/aws-cloudfront/lib/experimental/edge-function.ts:127]: NIT
```suggestion
  public grantInvokeCompositePrincipal(compositePrincipal: iam.CompositePrincipal): iam.Grant[] {
```

File: packages/aws-cdk-lib/aws-lambda/lib/function-base.ts [hunk 8]
```diff
@@ -102,6 +102,11 @@ export interface IFunction extends IResource, ec2.IConnectable, iam.IGrantable {
    */
   grantInvokeUrl(identity: iam.IGrantable): iam.Grant;
 
+  /**
+   * Grant multiple principals the ability to invoke this Lambda via CompositePrincipal
+   */
+  grantInvokeCompositePrincipal(principals: iam.CompositePrincipal): iam.Grant[];
```
Comment (vinayak-kukreja) [inline packages/aws-cdk-lib/aws-lambda/lib/function-base.ts:108]: NIT

```suggestion
  grantInvokeCompositePrincipal(principal: iam.CompositePrincipal): iam.Grant[];
```

File: packages/aws-cdk-lib/aws-lambda/lib/function-base.ts [hunk 9]
```diff
@@ -449,6 +454,13 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
     return grant;
   }
 
+  /**
+   * Grant multiple principals the ability to invoke this Lambda via CompositePrincipal
+   */
+  public grantInvokeCompositePrincipal(compositePrincipals: iam.CompositePrincipal): iam.Grant[] {
```
Comment (vinayak-kukreja) [inline packages/aws-cdk-lib/aws-lambda/lib/function-base.ts:460]: NIT
```suggestion
  public grantInvokeCompositePrincipal(compositePrincipal: iam.CompositePrincipal): iam.Grant[] {
```

**Review by vinayak-kukreja — CHANGES_REQUESTED** [review]
Hey @msambol , this is great. Thank you for opening this PR with us. 

I just have some NITs and please let me know if the `integ` tests are dry-run and I can run these for you.

File: packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.js.snapshot/lambda-permissions.assets.json (hunk 6 above)
Comment (msambol) [reply packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.js.snapshot/lambda-permissions.assets.json:2]: Hmm, I didn't dry-run..weird. I saw it deploy into my account.

Comment (vinayak-kukreja) [reply packages/@aws-cdk-testing/framework-integ/test/aws-lambda/test/integ.permissions.js.snapshot/lambda-permissions.assets.json:2]: That is fine. This happens in some instances. Thank you for confirming that you have deployed and tested these :)

**Pull request merged**